rclpy/test/test_talker_listener.py

```diff
@@ -12,126 +12,83 @@
-          .format(msg.header.frame_id, msg.child_frame_id, msg.transform.translation.x,
-                  msg.transform.rotation.x))
-    received_messages.append(msg)
-
-
-def listener_fixed_array_callback(msg, received_messages):
-    print('received: ({})'.format(msg.angular_velocity_covariance))
-    received_messages.append(msg)
-
-
-def listener_unbounded_array_callback(msg, received_messages):
-    print('received: ({})'.format(msg.intensities))
-    received_messages.append(msg)
-
-
-def listener_nested_array_callback(msg, received_messages):
-    print('received: ({})'.format([f.name for f in msg.fields]))
-    received_messages.append(msg)
-
-
-def launch_talker_listener(test_file, message_pkg, message_name, callback):
-    import rclpy
-    from rclpy.qos import qos_profile_default
-    import importlib
+def launch_talker_listener(message_pkg, message_name, nbmessages=5):
+    talker_file = 'talker.py'
+    listener_file = 'listener.py'
 
     launch_desc = LaunchDescriptor()
     launch_desc.add_process(
-        cmd=[sys.executable, os.path.join(this_dir, test_file)],
-        name='test_talker_listener__{}'.format(test_file)
+        cmd=[sys.executable,
+             os.path.join(this_dir, talker_file),
+             '-p', message_pkg,
+             '-m', message_name,
+             '-n', str(nbmessages)],
+        name='test_talker__{}'.format(message_name),
+        exit_handler=ignore_signal_exit_handler
+    )
+
+    launch_desc.add_process(
+        cmd=[sys.executable,
+             os.path.join(this_dir, listener_file),
+             '-p', message_pkg,
+             '-m', message_name,
+             '-n', str(nbmessages)],
+        name='test_listener__{}'.format(message_name),
+        exit_handler=primary_exit_handler
     )
     launcher = DefaultLauncher()
     launcher.add_launch_descriptor(launch_desc)
-    async_launcher = AsynchronousLauncher(launcher)
-    async_launcher.start()
-
-    rclpy.init([])
-
-    module = importlib.import_module(message_pkg + '.msg')
-    msg_mod = getattr(module, message_name)
-    assert msg_mod.__class__._TYPE_SUPPORT is not None
-
-    node = rclpy.create_node(test_file[:test_file.rfind('.')])
-
-    received_messages = []
-
-    chatter_callback = functools.partial(
-        callback, received_messages=received_messages)
-
-    # TODO(wjwwood): should the subscription object hang around?
-    node.create_subscription(
-        msg_mod,
-        test_file[:test_file.rfind('.')].replace('talker', 'chatter'),
-        chatter_callback,
-        qos_profile_default)
+    rc = launcher.launch()
 
     spin_count = 0
-    while len(received_messages) == 0 and spin_count < 10 and launcher.is_launch_running():
-        rclpy.spin_once(node)  # This will wait 1 second or until something has been done.
+    while spin_count < nbmessages and launcher.is_launch_running():
+        time.sleep(1)
         spin_count += 1
```

Comment:
But you can't just increment `msg_count` in the loop either. That semantic would be weird too. Currently `nbmessages` is more like "max-runtime-in_seconds".